File: src/net/azib/photos/cast/PhotoDirsSuggestionAdapter.java
package net.azib.photos.cast;

import android.app.Activity;
import android.content.res.Resources;
import android.widget.ArrayAdapter;
import android.widget.Filter;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.URL;
import java.net.URLConnection;
import java.util.ArrayList;
import java.util.List;

import static java.util.Collections.emptyList;

public class PhotoDirsSuggestionAdapter extends ArrayAdapter<String> {
//    protected static final String TAG = "SuggestionAdapter";
    private List<String> suggestions;

    public PhotoDirsSuggestionAdapter(Activity context) {
        super(context, android.R.layout.simple_dropdown_item_1line);
        suggestions = new ArrayList<>();
    }
 
    @Override
    public int getCount() {
        return suggestions.size();
    }
 
    @Override
    public String getItem(int index) {
        return suggestions.get(index);
    }

    public List<String> getSuggestions(String dir) {
        try {
            List<String> result = new ArrayList<>();
            Resources resources = getContext().getResources();
            URL js = new URL(resources.getString(R.string.photos_dirs_url) + "?dir=" + dir + "&accessToken=" + resources.getString(R.string.photos_dirs_access_token));
            URLConnection jc = js.openConnection();
            BufferedReader reader = new BufferedReader(new InputStreamReader(jc.getInputStream()));
            String line;
            while ((line = reader.readLine()) != null) result.add(line);
            return result;
        }
        catch (Exception e1) {
            return emptyList();
        }
    }
 
    @Override
    public Filter getFilter() {
        return new Filter() {
            @Override
            protected FilterResults performFiltering(CharSequence constraint) {
                FilterResults filterResults = new FilterResults();
                if (constraint != null) {
                    suggestions = getSuggestions(constraint.toString());
                    filterResults.values = suggestions;
                    filterResults.count = suggestions.size();
                }
                return filterResults;
            }

            @Override
            protected void publishResults(CharSequence contraint, FilterResults results) {
                if (results != null && results.count > 0) {
                    notifyDataSetChanged();
                }
                else {
                    notifyDataSetInvalidated();
                }
            }
        };
    }
 
}

File: src/net/azib/photos/cast/PhotoDirsSuggestionAdapter.kt
package net.azib.photos.cast

import android.app.Activity
import android.widget.ArrayAdapter
import android.widget.Filter
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.URL
import java.util.Collections.emptyList

class PhotoDirsSuggestionAdapter(context: Activity) : ArrayAdapter<String>(context, android.R.layout.simple_dropdown_item_1line) {
  private var suggestions: List<String> = emptyList()
  private val accessToken = context.getString(R.string.photos_dirs_access_token)
  private val url = context.getString(R.string.photos_dirs_url)

  override fun getCount() = suggestions.size
  override fun getItem(index: Int) = suggestions[index]

  fun getSuggestions(dir: String): List<String> {
    try {
      val url = URL("$url?dir=$dir&accessToken=$accessToken")
      BufferedReader(InputStreamReader(url.openStream())).useLines { return it.toList() }
    } catch (e: Exception) {
      return emptyList()
    }
  }

  override fun getFilter(): Filter {
    return object : Filter() {
      override fun performFiltering(constraint: CharSequence?): Filter.FilterResults {
        val results = Filter.FilterResults()
        if (constraint == null) return results
        suggestions = getSuggestions(constraint.toString())
        results.values = suggestions
        results.count = suggestions.size
        return results
      }

      override fun publishResults(constraint: CharSequence?, results: Filter.FilterResults) {
        if (results.count > 0)
          notifyDataSetChanged()
        else
          notifyDataSetInvalidated()
      }
    }
  }
}